lnwire/features_test.go

```diff
@@ -3,152 +3,284 @@ package lnwire
 	t.Parallel()
 
-	const first = "first"
+	fv := featureVector{
+		featureNames: testFeatureNames,
+	}
 
-	localFeatures := NewFeatureVector([]Feature{
-		{first, OptionalFlag},
-	})
-	localFeatures.SetFeatureFlag(first, RequiredFlag)
-	remoteFeatures := NewFeatureVector([]Feature{})
+	tests := []struct {
+		bits            []FeatureBit
+		expectedEncoded []byte
+	}{
+		{
+			bits:            nil,
+			expectedEncoded: []byte{0x00, 0x00},
+		},
+		{
+			bits:            []FeatureBit{2, 3, 7},
+			expectedEncoded: []byte{0x00, 0x01, 0x8C},
+		},
+		{
+			bits:            []FeatureBit{2, 3, 8},
+			expectedEncoded: []byte{0x00, 0x02, 0x01, 0x0C},
+		},
+	}
 
-	_, err := localFeatures.Compare(remoteFeatures)
-	if err == nil {
-		t.Fatalf("feature was set as required but error wasn't "+
-			"returned: %v", err)
+	for i, test := range tests {
+		fv.features = make(map[FeatureBit]bool)
+		for _, bit := range test.bits {
+			fv.Set(bit)
+		}
+
+		// Test that Encode produces the correct serialization.
+		buffer := new(bytes.Buffer)
+		err := fv.Encode(buffer)
+		if err != nil {
+			t.Errorf("Failed to encode feature vector in case %d: %v", i, err)
+			continue
+		}
+
+		encoded := buffer.Bytes()
+		if !bytes.Equal(encoded, test.expectedEncoded) {
+			t.Errorf("Wrong encoding in case %d: got %v, expected %v",
+				i, encoded, test.expectedEncoded)
+			continue
+		}
+
+		// Test that decoding then re-encoding produces the same result.
+		fv2 := &featureVector{
+			features: make(map[FeatureBit]bool),
+		}
+		err = fv2.Decode(bytes.NewReader(encoded))
+		if err != nil {
+			t.Errorf("Failed to decode feature vector in case %d: %v", i, err)
+			continue
+		}
+
+		buffer2 := new(bytes.Buffer)
+		err = fv2.Encode(buffer2)
+		if err != nil {
+			t.Errorf("Failed to re-encode feature vector in case %d: %v",
+				i, err)
+			continue
+		}
+
+		reencoded := buffer2.Bytes()
+		if !bytes.Equal(reencoded, test.expectedEncoded) {
+			t.Errorf("Wrong re-encoding in case %d: got %v, expected %v",
+				i, reencoded, test.expectedEncoded)
+		}
 	}
 }
 
-// TestDecodeEncodeFeaturesVector checks that feature vector might be
-// successfully encoded and decoded.
-func TestDecodeEncodeFeaturesVector(t *testing.T) {
+func TestFeatureVectorUnknownFeatures(t *testing.T) {
 	t.Parallel()
 
-	const first = "first"
-
-	f := NewFeatureVector([]Feature{
-		{first, OptionalFlag},
-	})
-
-	var b bytes.Buffer
-	if err := f.Encode(&b); err != nil {
-		t.Fatalf("error while encoding feature vector: %v", err)
+	fv := featureVector{
+		features:     make(map[FeatureBit]bool),
+		featureNames: testFeatureNames,
 	}
 
-	nf, err := NewFeatureVectorFromReader(&b)
-	if err != nil {
-		t.Fatalf("error while decoding feature vector: %v", err)
+	tests := []struct {
+		bits            []FeatureBit
+		expectedUnknown []FeatureBit
+	}{
+		{
+			bits:            nil,
+			expectedUnknown: nil,
+		},
+		{
+			bits:            []FeatureBit{0, 1, 2, 3, 4, 5, 6, 7},
```

Comment:
I think this case could use a comment explaining that only bits (as defined above) {0, 3, 4, 5} are known, therefore _only_ the two even bits (as odd is OK) should be flagged as unknown.